<a href="https://colab.research.google.com/github/drshahizan/Python-big-data/blob/main/Project/SIX/Final_project_hpdp.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **NYC Parking Ticket Analysis Using 3 Different Libraries.**

**Group Members:**

<table>
  <tr>
    <th>Name</th>
    <th>Matric</th>
  </tr>
  <tr>
    <th>LEE MING QI</th>
    <th>A20EC0064</th>
  </tr>

  <tr>
    <th>AMIRAH RAIHANAH BINTI ABDUL RAHIM</th>
    <th>A20EC0182</th>
  </tr>
    <tr>
    <th>SINGTHAI SRISOI</th>
    <th>A20EC0147</th>
  </tr>
    <tr>
    <th>NUR IRDINA ALIAH BINTI ABDUL WAHAB</th>
    <th>A20EC0115</th>
  </tr>
</table>


**About Project:**

In this project, we will analyze NYC Parking Ticket dataset by performing EDA using 3 different libraries. The libraries that will be use are Dask, Pyspark and Koalas.The dataset used can be retrieve from [Kaggle](https://www.kaggle.com/datasets/new-york-city/nyc-parking-tickets). This is a large dataset that contains 10 million rows and 51 columns with the size of 2GB. We will show each operation by using the 3 libraries stated and provide time taken needed to perform it. This will help you to evaluate which library can do it more effectively. 
This project is divided to 4 parts;


1.   Downloading Dataset
2.   Data Cleaning
3.   Data Visualization
4.   Data Questions and Answers
5.   Findings and Analysis






---


## **`Downloading the Dataset`**

The NYC Department of Finance collects data on every parking ticket issued in NYC. This data is made publicly available to aid in ticket resolution and to guide policymakers. Now, we will perform EDA by first downloading the dataset from Kaggle [here](https://www.kaggle.com/datasets/new-york-city/nyc-parking-tickets).

Download steps:
1. Open [dataset link](https://www.kaggle.com/datasets/new-york-city/nyc-parking-tickets) that will be directed to the NYC Parking Ticket dataset.
2. Download the dataset.
3. Upload dataset into Google Drive.
4. Get link of the dataset file in Google Drive.
5. Import the dataset retrieve from Google Drive.

In [ ]:
# Pull dataset from google drive
!pip install -U -q PyDrive
from pydrive.auth import GoogleAuth
from pydrive.drive import GoogleDrive
from google.colab import auth
from oauth2client.client import GoogleCredentials

# Authenticate and create the PyDrive client.
auth.authenticate_user()
gauth = GoogleAuth()
gauth.credentials = GoogleCredentials.get_application_default()
drive = GoogleDrive(gauth)
link = 'https://drive.google.com/file/d/1TSEOGseOzszVn342YESVHk7LyjX1Qmbd/view?usp=sharing'
# to get the id part of the file
id = link.split("/")[-2]
 
downloaded = drive.CreateFile({'id':id})
downloaded.GetContentFile('NYC_Parking_Ticket.csv') 



---

# **`Read Dataset Into Dataframe Of Each Library`**
Steps:
1. Install library using pip.
2. Import the library installed.
3. Read the csv file into dataframe.

In [ ]:
import pandas as pd
import numpy as np

**1. Dask**

In [ ]:
!python -m pip install "dask[complete]" # Install everything

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [ ]:
 #specify dtypes to prevent errors because the dataset is messy.
 dtypes={'Date First Observed': str,
    'Days Parking In Effect    ': str,
    'Double Parking Violation': str,
    'Feet From Curb': np.float32,
    'From Hours In Effect': str,
    'House Number': str,
    'Hydrant Violation': str,
    'Intersecting Street': str,
    'Issue Date': str,
    'Issuer Code': np.float32,
    'Issuer Command': str,
    'Issuer Precinct': np.float32,
    'Issuer Squad': str,
    'Issuing Agency': str,
    'Law Section': np.float32,
    'Meter Number': str,
    'No Standing or Stopping Violation': str,
    'Plate ID': str,
    'Plate Type': str,
    'Registration State': str,
    'Street Code1': np.int32,
    'Street Code2': np.int32,
    'Street Code3': np.int32,
    'Street Name': str,
    'Sub Division': str,
    'Summons Number': np.int32,
    'Time First Observed': str,
    'To Hours In Effect': str,
    'Unregistered Vehicle?': str,
    'Vehicle Body Type': str,
    'Vehicle Color': str,
    'Vehicle Expiration Date': str,
    'Vehicle Make': str,
    'Vehicle Year': np.float32,
    'Violation Code': np.int16,
    'Violation County': str,
    'Violation Description': str,
    'Violation In Front Of Or Opposite': str,
    'Violation Legal Code': str,
    'Violation Location': str,
    'Violation Post Code': str,
    'Violation Precinct': np.float32,
    'Violation Time': str
}

In [ ]:
import dask.dataframe as dd

In [ ]:
%%time
dfdask = dd.read_csv('NYC_Parking_Ticket.csv',dtype=dtypes)
dfdask.head()
       

CPU times: user 1.58 s, sys: 146 ms, total: 1.73 s
Wall time: 1.79 s


Summons Number Plate ID Registration State Plate Type  Issue Date  \
0      1283294138  GBB9093                 NY        PAS  08/04/2013   
1      1283294151  62416MB                 NY        COM  08/04/2013   
2      1283294163  78755JZ                 NY        COM  08/05/2013   
3      1283294175  63009MA                 NY        COM  08/05/2013   
4      1283294187  91648MC                 NY        COM  08/08/2013   

   Violation Code Vehicle Body Type Vehicle Make Issuing Agency  Street Code1  \
0              46              SUBN         AUDI              P         37250   
1              46               VAN         FORD              P         37290   
2              46               P-U        CHEVR              P         37030   
3              46               VAN         FORD              P         37270   
4              41              TRLR          GMC              P         37240   

   ...  Hydrant Violation  Double Parking Violation Latitude Longitude  \
0  ...                NaN                       NaN      NaN       NaN   
1  ...                NaN                       NaN      NaN       NaN   
2  ...                NaN                       NaN      NaN       NaN   
3  ...                NaN                       NaN      NaN       NaN   
4  ...                NaN                       NaN      NaN       NaN   

   Community Board  Community Council   Census Tract BIN BBL NTA  
0              NaN                 NaN           NaN NaN NaN NaN  
1              NaN                 NaN           NaN NaN NaN NaN  
2              NaN                 NaN           NaN NaN NaN NaN  
3              NaN                 NaN           NaN NaN NaN NaN  
4              NaN                 NaN           NaN NaN NaN NaN  

[5 rows x 51 columns]

**2. PySpark**

In [ ]:
!pip install pyspark

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 281.4/281.4 MB 5.4 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 199.7/199.7 KB 17.4 MB/s eta 0:00:00
  Created wheel for pyspark: filename=pyspark-3.3.1-py2.py3-none-any.whl size=281845512 sha256=fc118cce285163eb0b7dec64a8a9299a51e2ce82e51a2ed2f1aaa3aae84e54f2
  Stored in directory: /root/.cache/pip/wheels/43/dc/11/ec201cd671da62fa9c5cc77078235e40722170ceba231d7598
Successfully built pyspark


In [ ]:
from pyspark.sql import SparkSession
import pyspark.sql.functions as func
from pyspark.sql.types import StructType,StructField, StringType, IntegerType 
from pyspark.sql.types import ArrayType, DoubleType, BooleanType
from pyspark.sql.functions import col,array_contains
spark = SparkSession.builder.getOrCreate()

In [ ]:
%%time
dfs = spark.read.options(header = "true", \
  inferSchema = "true",) \
     .csv("NYC_Parking_Ticket.csv")
dfs.show(5)

+--------------+--------+------------------+----------+----------+--------------+-----------------+------------+--------------+------------+------------+------------+-----------------------+------------------+------------------+---------------+-----------+--------------+------------+--------------+-------------------+----------------+---------------------------------+------------+-----------+-------------------+-------------------+-----------+------------+--------------------+--------------------------+--------------------+------------------+-------------+---------------------+------------+------------+--------------+-------------------+---------------------+---------------------------------+-----------------+------------------------+--------+---------+---------------+------------------+------------+----+----+----+
|Summons Number|Plate ID|Registration State|Plate Type|Issue Date|Violation Code|Vehicle Body Type|Vehicle Make|Issuing Agency|Street Code1|Street Code2|Street Code3|Vehicle

**3. Koalas**

In [ ]:
!pip install koalas

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 390.8/390.8 KB 7.6 MB/s eta 0:00:00


In [ ]:
import databricks.koalas as ks

In [ ]:
%%time
dfk = ks.read_csv('NYC_Parking_Ticket.csv',dtype=dtypes)
dfk.head()

CPU times: user 4.23 s, sys: 698 ms, total: 4.92 s
Wall time: 1min 19s


Summons Number Plate ID Registration State Plate Type  Issue Date  Violation Code Vehicle Body Type Vehicle Make Issuing Agency  Street Code1  Street Code2  Street Code3 Vehicle Expiration Date Violation Location  Violation Precinct  Issuer Precinct  Issuer Code Issuer Command Issuer Squad Violation Time Time First Observed Violation County Violation In Front Of Or Opposite House Number Street Name Intersecting Street Date First Observed  Law Section Sub Division Violation Legal Code Days Parking In Effect     From Hours In Effect To Hours In Effect Vehicle Color Unregistered Vehicle?  Vehicle Year Meter Number  Feet From Curb Violation Post Code Violation Description No Standing or Stopping Violation Hydrant Violation Double Parking Violation Latitude Longitude Community Board Community Council  Census Tract   BIN   BBL   NTA
0      1283294138  GBB9093                 NY        PAS  08/04/2013              46              SUBN         AUDI              P         37250         13610         21190                20140831                 33                33.0             33.0     921043.0           0033         0000          0752A                None             None                                 F          712    W 175 ST                None                   0        408.0           F1                 None                    BBBBBBB                  ALL                ALL            GY                     0        2013.0            -             0.0                None                  None                              None              None                     None     None      None            None               None         None  None  None  None
1      1283294151  62416MB                 NY        COM  08/04/2013              46               VAN         FORD              P         37290         40404         40404                20140430                 33                33.0             33.0     921043.0           0033         0000          1240P                None               NY                                 O          201    W 177 ST                None                   0        408.0            C                 None                    BBBBBBB                  ALL                ALL            WH                     0        2012.0            -             0.0                None                  None                              None              None                     None     None      None            None               None         None  None  None  None
2      1283294163  78755JZ                 NY        COM  08/05/2013              46               P-U        CHEVR              P         37030         31190         13610                20140228                 33                33.0             33.0     921043.0           0033         0000          1243P                None               NY                                 O          520    W 163 ST                None                   0        408.0           F7                 None                    BBBBBBB                  ALL                ALL          None                     0           0.0            -             0.0                None                  None                              None              None                     None     None      None            None               None         None  None  None  None
3      1283294175  63009MA                 NY        COM  08/05/2013              46               VAN         FORD              P         37270         11710         12010                20141031                 33                33.0             33.0     921043.0           0033         0000          0232P                None               NY                                 O          517    W 176 ST                None                   0        408.0           F1                 None                    BBBBBBB                  ALL                ALL            WH                     0        2010.0            -             0

>**Findings ▶** 
All library has different way of importing the dataset and requires different functions to run it.
1. Dask 🕔
Dask is the fastest way to read the data by taking **1.79s** but we have to initialize its dtype. This is because there are a lot of mismatched data type in the dataset. It requires to import **dask.dataframe** in order to read it in dataframe. The function used is common like Pandas which is **read.csv()**.

>2. PySpark 🕔
Pyspark took a long time to read the dataset which is **1min 10s**. In pyspark, there is no need to initialize dtype because we can set **inferSchema = True** and it will assign the correct dtype. The function used is different from common Pandas which is **spark.read.options.**

>3. Koalas 🕔
Koalas took the **longest time** compared to Dask and PySpark which is **1min 10s**.
Most of Koalas funtions are the same with common Pandas which in this case we use **read.csv()** to read the data into dataframe.




---



---



# **`Data Preparation and Cleaning`**

The dataset contains 10 million rows and 51 columns of records which has a lot of missing data, mismatched data, unused columns and other mess that need to be clean. Before we can analyze the data, we must clean the data to obtain an accurate analysis. Now, we will perform data cleaning using 3 different libraries and evaluate which library does it more effectively.

# **`Step 1: Checking dataframe size to ensure all libraries read the same number of records.`**

**1. Dask**

In [ ]:
%%time
dfdask.shape

CPU times: user 8.43 ms, sys: 0 ns, total: 8.43 ms
Wall time: 9.06 ms


(Delayed('int-68bfe4bb-bf90-4109-aed6-d05c2fcf248e'), 51)

**2. PySpark**

In [ ]:
%%time
rows = dfs.count()
cols = len(dfs.columns)
print (rows, "," ,cols)


9100278 , 51
CPU times: user 45.7 ms, sys: 10.5 ms, total: 56.3 ms
Wall time: 6.47 s


**3. Koalas**

In [ ]:
%%time
dfk.shape

CPU times: user 59.7 ms, sys: 3.8 ms, total: 63.5 ms
Wall time: 9.49 s


(9100278, 51)

>**Findings ▶** 
All libraries managed to read the same no of rows and columns which are **9100278 rows and 51 columns**. To find the number and columns read by all libraries, **Dask and Koalas** use the same method which is **dataframe.shape**. Meanwhile, **Pyspark** is different and it **uses count to find rows and cols for columns**.

>1. Dask 🕔
Dask is the fastest way to find number of rows and columns by taking **9.06ms**.

>2. PySpark 🕔
PySpark comes next as it took **6.47s** to find number of rows and columns.

>3. Koalas 🕔
Koalas took the **longest time** compared to Dask and PySpark which is **9.49s**.




---



---



# **`Check and drop column that has more than 80% missing values.`**



This dataset contains a lot of missing and null values and certain columns have no values at all. Anyhow, dropping all missing values is not recommended as it will lose a lot of records. Some null values can be change to 0, for example column `Feet From Curb` because the attribute may not be collected but the parking ticket is still valid.In order to get a clean data, we must drop the columns and change the null values to get consistent result for our analysis.

Steps:
1. Check number of Nan values.
2. Drop columns that have more than 80% Nan values.
3. Fill in the rest of Nan values with 0.
4. Check if there is any Nan values left.

**1. Dask**

In [ ]:
%%time
dfdask.isnull().sum().compute()


CPU times: user 1min 15s, sys: 1.71 s, total: 1min 17s
Wall time: 1min 4s


Summons Number                             0
Plate ID                                3612
Registration State                         0
Plate Type                                 0
Issue Date                                 0
Violation Code                             0
Vehicle Body Type                     111377
Vehicle Make                           63273
Issuing Agency                             0
Street Code1                               0
Street Code2                               0
Street Code3                               0
Vehicle Expiration Date                    0
Violation Location                    671170
Violation Precinct                         0
Issuer Precinct                            0
Issuer Code                                0
Issuer Command                        654437
Issuer Squad                          654437
Violation Time                          2074
Time First Observed                  8080301
Violation County                      705205
Violation 

In [ ]:
%%time
dfdask=dfdask.drop(['Time First Observed','Violation Legal Code','Unregistered Vehicle?','No Standing or Stopping Violation', 'Hydrant Violation', 'Double Parking Violation','Latitude','Longitude','Community Board','Community Council ','Census Tract','BIN','BBL','NTA'], axis=1) 
dfdask.head()


CPU times: user 1.49 s, sys: 27.9 ms, total: 1.52 s
Wall time: 1.52 s


Summons Number Plate ID Registration State Plate Type  Issue Date  \
0      1283294138  GBB9093                 NY        PAS  08/04/2013   
1      1283294151  62416MB                 NY        COM  08/04/2013   
2      1283294163  78755JZ                 NY        COM  08/05/2013   
3      1283294175  63009MA                 NY        COM  08/05/2013   
4      1283294187  91648MC                 NY        COM  08/08/2013   

   Violation Code Vehicle Body Type Vehicle Make Issuing Agency  Street Code1  \
0              46              SUBN         AUDI              P         37250   
1              46               VAN         FORD              P         37290   
2              46               P-U        CHEVR              P         37030   
3              46               VAN         FORD              P         37270   
4              41              TRLR          GMC              P         37240   

   ...  Sub Division  Days Parking In Effect     From Hours In Effect  \
0  ...            F1                     BBBBBBB                  ALL   
1  ...             C                     BBBBBBB                  ALL   
2  ...            F7                     BBBBBBB                  ALL   
3  ...            F1                     BBBBBBB                  ALL   
4  ...            E1                     BBBBBBB                  ALL   

  To Hours In Effect  Vehicle Color  Vehicle Year  Meter Number  \
0                ALL             GY        2013.0             -   
1                ALL             WH        2012.0             -   
2                ALL            NaN           0.0             -   
3                ALL             WH        2010.0             -   
4                ALL             BR        2012.0             -   

  Feet From Curb Violation Post Code Violation Description  
0            0.0                 NaN                   NaN  
1            0.0                 NaN                   NaN  
2            0.0                 NaN                   NaN  
3            0.0                 NaN                   NaN  
4            0.0                 NaN                   NaN  

[5 rows x 37 columns]

In [ ]:
#Not all nan values need to be drop. Hence we fill the Nan values with 0.
%%time
dfdask = dfdask.fillna(0)


CPU times: user 12.7 ms, sys: 1.98 ms, total: 14.7 ms
Wall time: 14.8 ms


In [ ]:
dfdask.isnull().sum().compute()

Summons Number                       0
Plate ID                             0
Registration State                   0
Plate Type                           0
Issue Date                           0
Violation Code                       0
Vehicle Body Type                    0
Vehicle Make                         0
Issuing Agency                       0
Street Code1                         0
Street Code2                         0
Street Code3                         0
Vehicle Expiration Date              0
Violation Location                   0
Violation Precinct                   0
Issuer Precinct                      0
Issuer Code                          0
Issuer Command                       0
Issuer Squad                         0
Violation Time                       0
Violation County                     0
Violation In Front Of Or Opposite    0
House Number                         0
Street Name                          0
Intersecting Street                  0
Date First Observed      

**2.PySpark**

In [ ]:
%%time
from pyspark.sql.functions import isnan, when, count, col
dfs.select([count(when(isnan(c) | col(c).isNull(), c)).alias(c) for c in dfs.columns]).show()

+--------------+--------+------------------+----------+----------+--------------+-----------------+------------+--------------+------------+------------+------------+-----------------------+------------------+------------------+---------------+-----------+--------------+------------+--------------+-------------------+----------------+---------------------------------+------------+-----------+-------------------+-------------------+-----------+------------+--------------------+--------------------------+--------------------+------------------+-------------+---------------------+------------+------------+--------------+-------------------+---------------------+---------------------------------+-----------------+------------------------+--------+---------+---------------+------------------+------------+-------+-------+-------+
|Summons Number|Plate ID|Registration State|Plate Type|Issue Date|Violation Code|Vehicle Body Type|Vehicle Make|Issuing Agency|Street Code1|Street Code2|Street Code

In [ ]:
%%time
dfs.drop('Time First Observed','Violation Legal Code','Unregistered Vehicle?','No Standing or Stopping Violation', 'Hydrant Violation', 'Double Parking Violation','Latitude','Longitude','Community Board','Community Council ','Census Tract','BIN','BBL','NTA') \
    .printSchema()

root
 |-- Summons Number: long (nullable = true)
 |-- Plate ID: string (nullable = true)
 |-- Registration State: string (nullable = true)
 |-- Plate Type: string (nullable = true)
 |-- Issue Date: string (nullable = true)
 |-- Violation Code: integer (nullable = true)
 |-- Vehicle Body Type: string (nullable = true)
 |-- Vehicle Make: string (nullable = true)
 |-- Issuing Agency: string (nullable = true)
 |-- Street Code1: integer (nullable = true)
 |-- Street Code2: integer (nullable = true)
 |-- Street Code3: integer (nullable = true)
 |-- Vehicle Expiration Date: integer (nullable = true)
 |-- Violation Location: integer (nullable = true)
 |-- Violation Precinct: integer (nullable = true)
 |-- Issuer Precinct: integer (nullable = true)
 |-- Issuer Code: integer (nullable = true)
 |-- Issuer Command: string (nullable = true)
 |-- Issuer Squad: string (nullable = true)
 |-- Violation Time: string (nullable = true)
 |-- Violation County: string (nullable = true)
 |-- Violation In Fron

In [ ]:
#Not all nan values need to be drop. Hence we fill the Nan values with 0.
%%time
dfs=dfs.fillna(0)

CPU times: user 2.77 ms, sys: 0 ns, total: 2.77 ms
Wall time: 39 ms


In [ ]:
%%time
from pyspark.sql.functions import isnan, when, count, col
dfs.select([count(when(isnan(c) | col(c).isNull(), c)).alias(c) for c in dfs.columns]).show()

ConnectionRefusedError: ignored

**3.Koalas**

In [ ]:
%%time
dfk.isnull().sum()

CPU times: user 1.45 s, sys: 153 ms, total: 1.6 s
Wall time: 1min 12s


Summons Number                             0
Plate ID                                   0
Registration State                         0
Plate Type                                 0
Issue Date                                 0
Violation Code                             0
Vehicle Body Type                          0
Vehicle Make                               0
Issuing Agency                             0
Street Code1                               0
Street Code2                               0
Street Code3                               0
Vehicle Expiration Date                    0
Violation Location                         0
Violation Precinct                         0
Issuer Precinct                            0
Issuer Code                                0
Issuer Command                             0
Issuer Squad                               0
Violation Time                             0
Time First Observed                        0
Violation County                           0
Violation 

In [ ]:
%%time
dfk=dfk.drop(['Time First Observed','Violation Legal Code','Unregistered Vehicle?','No Standing or Stopping Violation', 'Hydrant Violation', 'Double Parking Violation','Latitude','Longitude','Community Board','Community Council ','Census Tract','BIN','BBL','NTA'], axis=1) 
dfk.head()

CPU times: user 50.6 ms, sys: 8.15 ms, total: 58.7 ms
Wall time: 171 ms


Summons Number Plate ID Registration State Plate Type  Issue Date  Violation Code Vehicle Body Type Vehicle Make Issuing Agency  Street Code1  Street Code2  Street Code3 Vehicle Expiration Date Violation Location  Violation Precinct  Issuer Precinct  Issuer Code Issuer Command Issuer Squad Violation Time Violation County Violation In Front Of Or Opposite House Number Street Name Intersecting Street Date First Observed  Law Section Sub Division Days Parking In Effect     From Hours In Effect To Hours In Effect Vehicle Color  Vehicle Year Meter Number  Feet From Curb Violation Post Code Violation Description
0      1283294138  GBB9093                 NY        PAS  08/04/2013              46              SUBN         AUDI              P         37250         13610         21190                20140831                 33                33.0             33.0     921043.0           0033         0000          0752A             None                                 F          712    W 175 ST                None                   0        408.0           F1                    BBBBBBB                  ALL                ALL            GY        2013.0            -             0.0                None                  None
1      1283294151  62416MB                 NY        COM  08/04/2013              46               VAN         FORD              P         37290         40404         40404                20140430                 33                33.0             33.0     921043.0           0033         0000          1240P               NY                                 O          201    W 177 ST                None                   0        408.0            C                    BBBBBBB                  ALL                ALL            WH        2012.0            -             0.0                None                  None
2      1283294163  78755JZ                 NY        COM  08/05/2013              46               P-U        CHEVR              P         37030         31190         13610                20140228                 33                33.0             33.0     921043.0           0033         0000          1243P               NY                                 O          520    W 163 ST                None                   0        408.0           F7                    BBBBBBB                  ALL                ALL          None           0.0            -             0.0                None                  None
3      1283294175  63009MA                 NY        COM  08/05/2013              46               VAN         FORD              P         37270         11710         12010                20141031                 33                33.0             33.0     921043.0           0033         0000          0232P               NY                                 O          517    W 176 ST                None                   0        408.0           F1                    BBBBBBB                  ALL                ALL            WH        2010.0            -             0.0                None                  None
4      1283294187  91648MC                 NY        COM  08/08/2013              41              TRLR          GMC              P         37240         12010         31190                       0                 33                33.0             33.0     921043.0           0033         0000          1239P               NY                                 F          590    W 174 ST                None                   0        408.0           E1                    BBBBBBB                  ALL                ALL            BR        2012.0            -             0.0                None                  None

In [ ]:
#Not all nan values need to be drop. Hence we fill the Nan values with 0.
%%time
dfk = dfk.fillna(0)

CPU times: user 400 ms, sys: 37.8 ms, total: 438 ms
Wall time: 2.33 s


In [ ]:
#dfk.isnull().sum()

Summons Number                       0
Plate ID                             0
Registration State                   0
Plate Type                           0
Issue Date                           0
Violation Code                       0
Vehicle Body Type                    0
Vehicle Make                         0
Issuing Agency                       0
Street Code1                         0
Street Code2                         0
Street Code3                         0
Vehicle Expiration Date              0
Violation Location                   0
Violation Precinct                   0
Issuer Precinct                      0
Issuer Code                          0
Issuer Command                       0
Issuer Squad                         0
Violation Time                       0
Violation County                     0
Violation In Front Of Or Opposite    0
House Number                         0
Street Name                          0
Intersecting Street                  0
Date First Observed      

>**Findings ▶** 
All libraries managed to read the same no of rows and columns which are **9100278 rows and 51 columns**. To find the number and columns read by all libraries, **Dask and Koalas** use the same method which is **dataframe.shape**. Meanwhile, **Pyspark** is different and it **uses count to find rows and cols for columns**.

>1. Dask 🕔
Dask is the fastest way to find number of rows and columns by taking **9.06ms**.

>2. PySpark 🕔
PySpark comes next as it took **6.47s** to find number of rows and columns.

>3. Koalas 🕔
Koalas took the **longest time** compared to Dask and PySpark which is **9.49s**.



---



---



# **`Convert column 'IssueDate' to datetime.`**

**1. Dask**

In [ ]:
%%time
dfdask['Issue Date'] = dd.to_datetime(dfdask['Issue Date'],format='%m/%d/%Y',errors='coerce')
dfdask.dtypes

CPU times: user 19.2 ms, sys: 891 µs, total: 20.1 ms
Wall time: 30.9 ms


Summons Number                                int32
Plate ID                                     object
Registration State                           object
Plate Type                                   object
Issue Date                           datetime64[ns]
Violation Code                                int16
Vehicle Body Type                            object
Vehicle Make                                 object
Issuing Agency                               object
Street Code1                                  int32
Street Code2                                  int32
Street Code3                                  int32
Vehicle Expiration Date                      object
Violation Location                           object
Violation Precinct                          float32
Issuer Precinct                             float32
Issuer Code                                 float32
Issuer Command                               object
Issuer Squad                                 object
Violation Ti

**2. PySpark**

In [ ]:
%%time
#dfs = dfs.withColumn("Issue Date", to_date(dfs.date_string))
#from pyspark.sql.functions import to_date, lit
#df.select("db_id", to_date(lit("1-12-2022"),"d-MM-yyyy")).show()
from pyspark.sql.types import StringType, DateType, FloatType
dfs = dfs.withColumn('Issue Date', func.to_date('Issue Date', 'dd/MM/yyyy'))
dfs.printSchema()


root
 |-- Summons Number: long (nullable = true)
 |-- Plate ID: string (nullable = true)
 |-- Registration State: string (nullable = true)
 |-- Plate Type: string (nullable = true)
 |-- Issue Date: date (nullable = true)
 |-- Violation Code: integer (nullable = true)
 |-- Vehicle Body Type: string (nullable = true)
 |-- Vehicle Make: string (nullable = true)
 |-- Issuing Agency: string (nullable = true)
 |-- Street Code1: integer (nullable = true)
 |-- Street Code2: integer (nullable = true)
 |-- Street Code3: integer (nullable = true)
 |-- Vehicle Expiration Date: integer (nullable = true)
 |-- Violation Location: integer (nullable = true)
 |-- Violation Precinct: integer (nullable = true)
 |-- Issuer Precinct: integer (nullable = true)
 |-- Issuer Code: integer (nullable = true)
 |-- Issuer Command: string (nullable = true)
 |-- Issuer Squad: string (nullable = true)
 |-- Violation Time: string (nullable = true)
 |-- Time First Observed: string (nullable = true)
 |-- Violation County

**3. Koalas**

In [ ]:
%%time
dfk['Issue Date'] = ks.to_datetime(dfk['Issue Date'],format='%m/%d/%Y',errors='coerce')
dfk.dtypes

/usr/local/lib/python3.8/dist-packages/pyspark/sql/pandas/functions.py:394: UserWarning: In Python 3.6+ and Spark 3.0+, it is preferred to specify type hints for pandas UDF instead of specifying pandas UDF type which will be deprecated in the future releases. See SPARK-28264 for more details.
  warnings.warn(


CPU times: user 153 ms, sys: 17.7 ms, total: 170 ms
Wall time: 1.13 s


Summons Number                            int32
Plate ID                                 object
Registration State                       object
Plate Type                               object
Issue Date                           datetime64
Violation Code                            int32
Vehicle Body Type                        object
Vehicle Make                             object
Issuing Agency                           object
Street Code1                              int32
Street Code2                              int32
Street Code3                              int32
Vehicle Expiration Date                  object
Violation Location                       object
Violation Precinct                      float32
Issuer Precinct                         float32
Issuer Code                             float32
Issuer Command                           object
Issuer Squad                             object
Violation Time                           object
Violation County                        



---



---



**1. Dask**

In [ ]:
%%time
dfdask.drop_duplicates(subset = ['Summons Number'], inplace = True)
dfdask.head()


CPU times: user 2.27 s, sys: 47.8 ms, total: 2.32 s
Wall time: 2.33 s


Summons Number Plate ID Registration State Plate Type Issue Date  \
0      1283294138  GBB9093                 NY        PAS 2013-08-04   
1      1283294151  62416MB                 NY        COM 2013-08-04   
2      1283294163  78755JZ                 NY        COM 2013-08-05   
3      1283294175  63009MA                 NY        COM 2013-08-05   
4      1283294187  91648MC                 NY        COM 2013-08-08   

   Violation Code Vehicle Body Type Vehicle Make Issuing Agency  Street Code1  \
0              46              SUBN         AUDI              P         37250   
1              46               VAN         FORD              P         37290   
2              46               P-U        CHEVR              P         37030   
3              46               VAN         FORD              P         37270   
4              41              TRLR          GMC              P         37240   

   ...  Sub Division  Days Parking In Effect     From Hours In Effect  \
0  ...            F1                     BBBBBBB                  ALL   
1  ...             C                     BBBBBBB                  ALL   
2  ...            F7                     BBBBBBB                  ALL   
3  ...            F1                     BBBBBBB                  ALL   
4  ...            E1                     BBBBBBB                  ALL   

  To Hours In Effect  Vehicle Color  Vehicle Year  Meter Number  \
0                ALL             GY        2013.0             -   
1                ALL             WH        2012.0             -   
2                ALL              0           0.0             -   
3                ALL             WH        2010.0             -   
4                ALL             BR        2012.0             -   

  Feet From Curb Violation Post Code Violation Description  
0            0.0                   0                     0  
1            0.0                   0                     0  
2            0.0                   0                     0  
3            0.0                   0                     0  
4            0.0                   0                     0  

[5 rows x 37 columns]

**2. PySpark**

In [ ]:
#dfs.drop_duplicates(subset = ['Summons Number'], inplace = True)
%%time
dfs.dropDuplicates(['Summons Number'])
dfs.show(5)

+--------------+--------+------------------+----------+----------+--------------+-----------------+------------+--------------+------------+------------+------------+-----------------------+------------------+------------------+---------------+-----------+--------------+------------+--------------+-------------------+----------------+---------------------------------+------------+-----------+-------------------+-------------------+-----------+------------+--------------------+--------------------------+--------------------+------------------+-------------+---------------------+------------+------------+--------------+-------------------+---------------------+---------------------------------+-----------------+------------------------+--------+---------+---------------+------------------+------------+----+----+----+
|Summons Number|Plate ID|Registration State|Plate Type|Issue Date|Violation Code|Vehicle Body Type|Vehicle Make|Issuing Agency|Street Code1|Street Code2|Street Code3|Vehicle

**3. Koalas**

In [ ]:
%%time
dfk.drop_duplicates(subset = ['Summons Number'], inplace = True)
dfk.head()


CPU times: user 366 ms, sys: 35.8 ms, total: 402 ms
Wall time: 2.55 s


Py4JJavaError: ignored

----------------------------------------
Exception happened during processing of request from ('127.0.0.1', 56754)
ERROR:root:Exception while sending command.
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/dist-packages/py4j/clientserver.py", line 516, in send_command
    raise Py4JNetworkError("Answer from Java side is empty")
py4j.protocol.Py4JNetworkError: Answer from Java side is empty

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.8/dist-packages/py4j/java_gateway.py", line 1038, in send_command
    response = connection.send_command(command)
  File "/usr/local/lib/python3.8/dist-packages/py4j/clientserver.py", line 539, in send_command
    raise Py4JNetworkError(
py4j.protocol.Py4JNetworkError: Error while sending or receiving
ERROR:root:Exception while sending command.
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/dist-packages/py4j/clientserver.py"

Py4JError: ignored

# Exploratory Analysis and Visualization





> Instructions (delete this cell)

* Compute the mean, sum, range and other 
interesting statistics for numeric columns
* Explore distributions of numeric columns using histograms etc.
* Explore relationship between columns using scatter plots, bar charts etc.
* Make a note of interesting insights from the exploratory analysis



**Pandas**

**Dask**

**Polars**

# Asking and Answering Questions

> Instructions (delete this cell)

* Ask at least 5 interesting questions about your dataset
* Answer the questions either by computing the results using Numpy/Pandas or by plotting graphs using Matplotlib/Seaborn
* Create new columns, merge multiple dataset and perform grouping/aggregation wherever necessary
* Wherever you're using a library function from Pandas/Numpy/Matplotlib etc. explain briefly what it does

### 1. Dask

a. What is the most common plate type?

In [ ]:
%%time
result = dfdask.groupby("Plate Type").agg({"Plate Type" : "count"}).compute()
result.index.name = 'Name'
result.sort_values(by="Plate Type",ascending=False,inplace=True)
result.head(5)

CPU times: user 1min 27s, sys: 1.93 s, total: 1min 29s
Wall time: 1min 22s


Plate Type
Name            
PAS      6511094
COM      1887869
OMT       235273
SRF        96764
OMS        75041

In [ ]:
dfdask.head(5)

Summons Number Plate ID Registration State Plate Type Issue Date  \
0      1283294138  GBB9093                 NY        PAS 2013-08-04   
1      1283294151  62416MB                 NY        COM 2013-08-04   
2      1283294163  78755JZ                 NY        COM 2013-08-05   
3      1283294175  63009MA                 NY        COM 2013-08-05   
4      1283294187  91648MC                 NY        COM 2013-08-08   

   Violation Code Vehicle Body Type Vehicle Make Issuing Agency  Street Code1  \
0              46              SUBN         AUDI              P         37250   
1              46               VAN         FORD              P         37290   
2              46               P-U        CHEVR              P         37030   
3              46               VAN         FORD              P         37270   
4              41              TRLR          GMC              P         37240   

   ...  Sub Division  Days Parking In Effect     From Hours In Effect  \
0  ...            F1                     BBBBBBB                  ALL   
1  ...             C                     BBBBBBB                  ALL   
2  ...            F7                     BBBBBBB                  ALL   
3  ...            F1                     BBBBBBB                  ALL   
4  ...            E1                     BBBBBBB                  ALL   

  To Hours In Effect  Vehicle Color  Vehicle Year  Meter Number  \
0                ALL             GY        2013.0             -   
1                ALL             WH        2012.0             -   
2                ALL              0           0.0             -   
3                ALL             WH        2010.0             -   
4                ALL             BR        2012.0             -   

  Feet From Curb Violation Post Code Violation Description  
0            0.0                   0                     0  
1            0.0                   0                     0  
2            0.0                   0                     0  
3            0.0                   0                     0  
4            0.0                   0                     0  

[5 rows x 37 columns]

b. What registeration stae has the highest rate?

In [ ]:
%%time
result = dfdask.groupby("Registration State").agg({"Plate Type" : "count"}).compute()
result.index.name = 'Name'
result.sort_values(by="Plate Type",ascending=False,inplace=True)
result.head(5)

CPU times: user 1min 29s, sys: 1.65 s, total: 1min 30s
Wall time: 1min 19s


Plate Type
Name            
NY       7029804
NJ        878677
PA        225760
CT        136973
FL        111887

c. What is the most common vehicle body type?

d. What is the most common vehicle colour?

e. What is the most common state?

In [ ]:
dfdask.head()

Summons Number Plate ID Registration State Plate Type Issue Date  \
0      1283294138  GBB9093                 NY        PAS 2013-08-04   
1      1283294151  62416MB                 NY        COM 2013-08-04   
2      1283294163  78755JZ                 NY        COM 2013-08-05   
3      1283294175  63009MA                 NY        COM 2013-08-05   
4      1283294187  91648MC                 NY        COM 2013-08-08   

   Violation Code Vehicle Body Type Vehicle Make Issuing Agency  Street Code1  \
0              46              SUBN         AUDI              P         37250   
1              46               VAN         FORD              P         37290   
2              46               P-U        CHEVR              P         37030   
3              46               VAN         FORD              P         37270   
4              41              TRLR          GMC              P         37240   

   ...  Sub Division  Days Parking In Effect     From Hours In Effect  \
0  ...            F1                     BBBBBBB                  ALL   
1  ...             C                     BBBBBBB                  ALL   
2  ...            F7                     BBBBBBB                  ALL   
3  ...            F1                     BBBBBBB                  ALL   
4  ...            E1                     BBBBBBB                  ALL   

  To Hours In Effect  Vehicle Color  Vehicle Year  Meter Number  \
0                ALL             GY        2013.0             -   
1                ALL             WH        2012.0             -   
2                ALL              0           0.0             -   
3                ALL             WH        2010.0             -   
4                ALL             BR        2012.0             -   

  Feet From Curb Violation Post Code Violation Description  
0            0.0                   0                     0  
1            0.0                   0                     0  
2            0.0                   0                     0  
3            0.0                   0                     0  
4            0.0                   0                     0  

[5 rows x 37 columns]

In [ ]:
result = dfdask.groupby('Plate Type').agg({'Plate Type' : 'count'}).compute()


In [ ]:
result.

SyntaxError: ignored

### 2. PySpark

a. What is the most common plate type?

b. What registeration stae has the highest rate?

c. What is the most common vehicle body type?

d. What is the most common vehicle colour?

e. What is the most common state?

In [ ]:
dfs.select('*').show(5)

### 3. Koalas

a. What is the most common plate type?

In [ ]:
%%time
dfk = ks.read_csv('NYC_Parking_Ticket.csv')
dfk.head()

In [ ]:
%%time
dfk.groupby('Plate Type').agg({'Plate Type' : 'count'})

b. What registeration state has the highest rate?

In [ ]:
%%time
dfk.groupby('Registration State').count()

c. What is the most common vehicle body type?

In [ ]:
%%time
dfk.groupby('Vehicle Body Type').count()

d. What is the most common vehicle colour?

In [ ]:
%%time
dfk.groupby('').count()

e. What is the most common state?

In [ ]:
%%time
dfk.groupby('').count()

# Inferences and Conclusion

**TODO** - Write some explanation here: a summary of all the inferences drawn from the analysis, and any conclusions you may have drawn by answering various questions.

# References and Future Work

**TODO** - Write some explanation here: ideas for future projects using this dataset, and links to resources you found useful.